cudaaligner/src/hirschberg_myers_gpu.cu

```diff
@@ -505,80 +483,159 @@ __device__ void hirschberg_myers_single_char_warp(int8_t*& path, int32_t* path_l
             ++path;
             --t;
         }
-        *path_length += target_end - target_begin;
     }
 }
 
-struct query_target_range
+template <typename T>
+struct parallel_warp_shared_stack_state
 {
-    char const* query_begin  = nullptr;
-    char const* query_end    = nullptr;
-    char const* target_begin = nullptr;
-    char const* target_end   = nullptr;
+    uint32_t mutex_;
+    int32_t active_warps_;
+    T* buffer_begin_;
+    T* cur_end_;
+    T* buffer_end_;
 };
 
 template <typename T>
-class warp_shared_stack
+class parallel_warp_shared_stack
 {
 public:
-    __device__ warp_shared_stack(char* buffer_begin, char* buffer_end)
-        : buffer_begin_(buffer_begin), cur_end_(buffer_begin), buffer_end_(buffer_end)
+    __device__ parallel_warp_shared_stack(parallel_warp_shared_stack_state<T>* data, T* buffer_begin, T* buffer_end)
+        : data_(data)
     {
-        assert(buffer_begin_ < buffer_end_);
+        assert(buffer_begin < buffer_end);
+        if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0)
+        {
+            data_->mutex_        = 0;
+            data_->active_warps_ = 0;
+            data_->buffer_begin_ = buffer_begin;
+            data_->cur_end_      = buffer_begin;
+            data_->buffer_end_   = buffer_end;
+        }
+        __syncthreads();
+        if (threadIdx.x % warp_size == 0)
+        {
+            atomicAdd(&data_->active_warps_, 1);
+        }
+        __syncthreads();
     }
 
-    __device__ bool inline push(T const& t, unsigned warp_mask = 0xffff'ffff)
+    __device__ bool inline push(T const& t)
     {
-        if (buffer_end_ - cur_end_ >= sizeof(T))
+        lock_mutex_high_priority();
+        bool success = false;
+        __syncwarp();
+
+        if (data_->buffer_end_ - data_->cur_end_ >= 1)
         {
-            __syncwarp(warp_mask);
-            if (threadIdx.x == 0)
+            __syncwarp();
+            if (threadIdx.x % warp_size == 0)
             {
-                std::memcpy(cur_end_, &t, sizeof(T));
+                *(data_->cur_end_) = t;
+                ++(data_->cur_end_);
             }
-            __syncwarp(warp_mask);
-            cur_end_ += sizeof(T);
-            return true;
+            success = true;
         }
         else
         {
-            if (threadIdx.x == 0)
+            if (threadIdx.x % warp_size == 0)
             {
                 printf("ERROR: stack full!");
             }
-            return false;
         }
+        __syncwarp();
+        release_mutex();
+        return success;
     }
 
-    __device__ inline void pop()
+    __device__ inline bool pop(T& element)
     {
-        assert(cur_end_ > buffer_begin_);
-        if (cur_end_ - sizeof(T) >= buffer_begin_)
-            cur_end_ -= sizeof(T);
+        lock_mutex();
+        if (threadIdx.x % warp_size == 0)
+        {
+            assert(data_->active_warps_ > 0);
+            --(data_->active_warps_);
```

Comment:
I'm not sure I follow this - could you explain why the active warp count is decremented at the beginning of the pop command, and then incremented later in the condition `if (data_->buffer_begin_ < data_->cur_end_)`?

Reply:
hmm maybe I understand now. so the `pop` command spins till it gets more data. when it gets data, it knows there will be an active warp to process it. and if new data never comes along, then it ends. did I sort of get it?

Reply:
Yes, I think you got it. Here's my explanation just to make sure:
The `active_warps_` variable is used to determine when all work is done and we can exit. Simply checking for an empty queue would be wrong, because a warp working on a task could fill the queue with new tasks again. Therefore, to check for completion, we need to check that the queue is empty and that there are no warps working on any task anymore. `active_warps_` counts the number of warps that are currently working on a task. If a warp calls `pop()` the warp has obviously no task to work on anymore, so it is not active anymore. When it then gets an task item from the queue (potentially after a bit of spinning in the `while(1)` loop), it will be active again.

This line https://github.com/clara-genomics/ClaraGenomicsAnalysis/blob/86a22a94ac5eb46740674a2aab668b4d53e4a25f/cudaaligner/src/hirschberg_myers_gpu.cu#L576 is the check which let's `pop()` return false (meaning "everything done!") and will exit the work loop in the caller.